src/main/scala/com/bazaarvoice/sswf/model/StepsHistory.scala

```diff
@@ -0,0 +1,171 @@
+            val stepStart: DateTime = new DateTime(startTimes(id))
+
+            val enum: StepEnum = enumFromName(id)
+            completedStepCounts = completedStepCounts + (enum -> (completedStepCounts(enum) + 1))
+            Some(StepEvent[StepEnum](eventId, Left(enum), StepResult.serialize(result), COMPLETE, eventStart, Some(endTime), new Duration(stepStart, endTime)))
+          case ActivityTaskTimedOut       =>
+            steps.put(h.getEventId, h)
+            val attributes: ActivityTaskTimedOutEventAttributes = h.getActivityTaskTimedOutEventAttributes
+            val eventId: Long = attributes.getScheduledEventId
+            val id: String = steps(eventId).getActivityTaskScheduledEventAttributes.getActivityId
+            val timeoutType: String = attributes.getTimeoutType
+            val result: TimedOut = TimedOut(timeoutType)
+            val eventStart =
+              if (timeoutType == "SCHEDULE_TO_START") {
+                new DateTime(steps(eventId).getEventTimestamp)
+              } else {
+                new DateTime(steps(attributes.getStartedEventId).getEventTimestamp)
+              }
+            val endTime: DateTime = new DateTime(h.getEventTimestamp.getTime)
+            val stepStart: DateTime = new DateTime(startTimes(id))
+            Some(StepEvent[StepEnum](eventId, Left(enumFromName(id)), StepResult.serialize(result), TIMED_OUT, eventStart, Some(endTime), new Duration(stepStart, endTime)))
+          case ActivityTaskFailed         =>
+            steps.put(h.getEventId, h)
+            val attributes: ActivityTaskFailedEventAttributes = h.getActivityTaskFailedEventAttributes
+            val eventId: Long = attributes.getScheduledEventId
+            val id: String = steps(eventId).getActivityTaskScheduledEventAttributes.getActivityId
+            val endTime: DateTime = new DateTime(h.getEventTimestamp.getTime)
+            val stepStart: DateTime = new DateTime(startTimes(id))
+            val result = Failed(s"${attributes.getReason}:${attributes.getDetails}")
+            Some(StepEvent[StepEnum](eventId, Left(enumFromName(id)), StepResult.serialize(result), FAILED, startTime, Some(endTime), new Duration(stepStart, endTime)))
+          case TimerStarted               =>
+            val attributes = h.getTimerStartedEventAttributes
+            val timerId = attributes.getTimerId
+            startedTimers.put(timerId, h)
+            None
+          case TimerFired                 =>
+            val attributes = h.getTimerFiredEventAttributes
+            val timerId = attributes.getTimerId
+            firedTimers.add(timerId)
+            None
+          case TimerCanceled              =>
+            val attributes = h.getTimerCanceledEventAttributes
+            val timerId = attributes.getTimerId
+            cancelledTimers.add(timerId)
+            None
+          case _                          =>
+            None
```

Comment:
there's a whole menagerie of other event types, and we just don't care about them.